File: 2025/python/day01.py
import operator

INPUT_PATH = "../input/day01.txt"

DIAL_START = 50


def part_1(instructions: list[str]) -> int:
    current_number = DIAL_START
    zero_count = 0

    for instruction in instructions:
        direction = instruction[0]
        num_clicks = int(instruction[1:])

        current_number = rotate(current_number, num_clicks, direction)

        if current_number == 0:
            zero_count += 1

    return zero_count


def rotate(current_number: int, num_clicks: int, direction: str) -> int:
    operation = {"L": operator.sub, "R": operator.add}[direction]
    difference = num_clicks % 100  # Ignore full circles
    new_number = operation(current_number, difference)

    # Adjust if we've gone past 0 or 99
    if new_number < 0:
        new_number += 100

    elif new_number > 99:
        new_number -= 100

    assert 0 <= new_number <= 99

    return new_number


if __name__ == "__main__":
    with open(INPUT_PATH, "r", encoding="utf-8") as f:
        instructions = f.readlines()

    print("Part 1:")
    print(part_1(instructions))

File: 2025/rust/src/day01.rs
use crate::helpers;

pub fn part_1() {
    if let Ok(lines) = helpers::read_lines("../input/day01.txt") {
        let mut zero_count = 0;
        let mut current_number = 50;

        for line in lines.map_while(Result::ok) {
            let (direction, num_clicks) = parse_line(line);

            current_number = rotate(current_number, direction, num_clicks);

            if current_number == 0 {
                zero_count += 1;
            }
        }

        println!("{}", zero_count)
    }
}

fn rotate(mut current_number: i32, direction: char, num_clicks: i32) -> i32 {
    let difference = num_clicks % 100;

    if direction == 'L' {
        current_number -= difference;
    } else if direction == 'R' {
        current_number += difference;
    } else {
        panic!("Unhandled direction");
    }

    if current_number < 0 {
        current_number += 100;
    } else if current_number > 99 {
        current_number -= 100;
    }

    current_number
}

fn parse_line(line: String) -> (char, i32) {
    let mut chars = line.chars();

    let direction: char = chars.next().expect("Can't parse direction");
    let num_clicks: i32 = chars
        .collect::<String>()
        .parse::<i32>()
        .expect("Can't parse num_clicks");

    (direction, num_clicks)
}
